### ISSUE TITLE
Refactor repositories to improve CRUD operations and querying support

### ISSUE THREAD

**Issue by bizz84** [bizz84/starter_architecture_flutter_firebase#115]
Currently, there is no way to let Firestore generate the document ID when adding new documents.

This is because the `FirestoreRepository` only uses `CollectionReference.set` (which requires the ID to be known beforehand), rather than `add` or `update`.

In addition, it would be good to expose methods that return a `Query` (as this will be needed for #114), but the current `FirestoreDataSource` class makes this difficult.

In summary, this issue is about overall improvements to the data layer and removing some of the old logic, which was not flexible enough.
